core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionValidator.scala

```diff
@@ -0,0 +1,50 @@
+package org.broadinstitute.dsde.rawls.expressions
+
+import org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport
+import org.broadinstitute.dsde.rawls.jobexec.MethodConfigResolver
+import org.broadinstitute.dsde.rawls.model.{AttributeString, ErrorReport, MethodConfiguration, ValidatedMethodConfiguration}
+import spray.http.StatusCodes
+
+import scala.util.Try
+
+object ExpressionValidator {
+  // inputsToParse is parameterized in order to validate in the presence or absence of the associated Method
+  // presence: inputs which are both empty and optional are pre-validated, so they are skipped here
+  // absence: validate all inputs normally
+
+  private[expressions] def validateAndParse(methodConfiguration: MethodConfiguration, inputsToParse: Map[String, AttributeString], parser: SlickExpressionParser): ValidatedMethodConfiguration = {
+    val (emptyOutputs, outputsToParse) = methodConfiguration.outputs.partition { case (_, expr) => expr.value.isEmpty }
+
+    val parsed = ExpressionParser.parseMCExpressions(inputsToParse, outputsToParse, parser)
+
+    // empty output expressions are also valid
+    val validatedOutputs = emptyOutputs.keys.toSeq ++ parsed.validOutputs
+
+    ValidatedMethodConfiguration(methodConfiguration, parsed.validInputs, parsed.invalidInputs, validatedOutputs, parsed.invalidOutputs)
+  }
+
+  // validate a MC without retrieving the Method from the Repo: assume all inputs are required
```

Comment:
Rob got tripped up on this too so maybe some rewording is in order.

There are two different validators here.

`validateAndParseMCExpressions` for MC /validate, create, save, get, import, etc.
It does not know about the Method because it did not query Agora, so it cannot know that any inputs are optional.  It doesn't fail anything, only reports which are invalid.

`validateExpressionsForSubmission` is for Create Submission and Validate Submission.  It does know about optional inputs because it has the Agora Method.  It does validate optional inputs if they are non-empty, and it does fail when any inputs/outputs are invalid.